soco/discovery.py

```diff
@@ -77,14 +92,34 @@ def discover(timeout=1, include_invisible=False, interface_addr=None):
         except socket.error:
             raise ValueError("{0} is not a valid IP address string".format(
                 interface_addr))
-        _sock.setsockopt(
-            socket.IPPROTO_IP, socket.IP_MULTICAST_IF, address)
-
-    # Send a few times. UDP is unreliable
-    _LOG.info("Sending discovery packets")
-    _sock.sendto(really_utf8(PLAYER_SEARCH), (MCAST_GRP, MCAST_PORT))
-    _sock.sendto(really_utf8(PLAYER_SEARCH), (MCAST_GRP, MCAST_PORT))
-    _sock.sendto(really_utf8(PLAYER_SEARCH), (MCAST_GRP, MCAST_PORT))
+        _sockets[0] = create_socket(interface_addr)
+    else:
+        # Find the local network address using a couple of different methods.
+        # Create a socket for each unique address found, and one for the
+        # default multicast address
+        addresses = set()
+        try:
+            addresses.add(socket.gethostbyname(socket.gethostname()))
+        except socket.error:
+            pass
+        try:
+            addresses.add(socket.gethostbyname(socket.getfqdn()))
+        except socket.error:
+            pass
+        for address in addresses:
+            _sockets.append(create_socket(address))
+        # Add a socket using the system default address
+        _sockets.append(create_socket(None))
+
+    _LOG.info(
+        "Sending discovery packets on default interface and %s",
+        list(addresses)
+    )
+    for _sock in _sockets:
```

Comment:
No major reason (other than removing copy/pasted code) - but looping each interface, sending each in turn before repeating would generally be seen as better as it gives the last one in the list its first message a little earlier, before repeating the send to the first one. Also if there was a problem with a single interface that it "blocks for a second because of some error", then you wouldn't block for a long time before you send to the next interface.